lib/SILOptimizer/Differentiation/PullbackCloner.cpp

```diff
@@ -362,6 +380,93 @@ class PullbackCloner::Implementation final
     return result;
   }
 
+  /// Used to materialize an aggregate adjoint directly, if
+  /// `isAdjointPiecewiseMaterializable` returned false.
+  SILValue materializeAggregateAdjointDirect(AdjointValue val,
+                                             SILLocation loc) {
+    SILValue result;
+    auto *resultAlloc = builder.createAllocStack(loc, val.getType());
+    builder.emitZeroIntoBuffer(loc, resultAlloc, IsInitialization);
+
+    if (auto *tupTy = val.getSwiftType()->getAs<TupleType>()) {
+      for (auto fieldIndex : range(val.getNumAggregateElements())) {
+        auto adjField = val.getAggregateElement(fieldIndex);
+        // No need to materialize zero field adjoints when we
+        // have already materialized a zero aggregate adjoint
+        // for the aggregate type containing the field.
+        if (adjField.getKind() != AdjointValueKind::Zero) {
+          auto eltTy = SILType::getPrimitiveAddressType(
+              tupTy->getElementType(fieldIndex)->getCanonicalType());
+          auto lhsAdjEltBuf =
+              builder
+                  .createTupleElementAddr(loc, resultAlloc, fieldIndex, eltTy)
+                  ->getResult(0);
+          auto rhsAdjEltBuf =
+              builder.createAllocStack(loc, adjField.getType())->getResult(0);
+          materializeAdjointIndirect(adjField, rhsAdjEltBuf, loc);
+
+          // lhsAdjEltBuf += rhsAdjEltBuf
```

Comment:
I think it's up to the caller to decide what to do with adjoint value. Can you fold these two into adjoint generation of `tuple_extract` and `struct_extract` correspondingly?

Reply:
To be explicit. What I'm suggesting:

For `y = struct_extract x, #field` the adjoint generation would be:
 1. You take adjoint of 'x'
 1. You take adjoint of 'y'
 1. You take address of #field in x'
 1. You do x'.#field += y'